File: scripts/fix-orientation.py
#!/usr/bin/env python3
"""
Fix image orientation using EXIF data.
Uses ImageOps.exif_transpose() to properly rotate images based on their EXIF orientation tag.
"""

import os
from pathlib import Path
from PIL import Image, ImageOps

def fix_image_orientation(image_path):
    """Fix the orientation of a single image using EXIF data."""
    try:
        with Image.open(image_path) as img:
            # Get original format
            original_format = img.format

            # Check if image has EXIF orientation data
            exif = img.getexif()
            orientation = exif.get(274) if exif else None  # 274 is the orientation tag

            if orientation and orientation != 1:
                # Apply EXIF transpose to fix orientation
                fixed_img = ImageOps.exif_transpose(img)

                # Save with same quality settings
                if original_format == 'JPEG':
                    fixed_img.save(image_path, 'JPEG', quality=95, exif=b'')
                else:
                    fixed_img.save(image_path)

                print(f"Fixed: {image_path} (orientation was {orientation})")
                return True
            else:
                print(f"Skipped: {image_path} (no rotation needed)")
                return False

    except Exception as e:
        print(f"Error processing {image_path}: {e}")
        return False

def process_directory(directory):
    """Process all images in a directory."""
    image_extensions = {'.jpg', '.jpeg', '.png'}
    fixed_count = 0
    skipped_count = 0
    error_count = 0

    for root, dirs, files in os.walk(directory):
        for file in files:
            ext = Path(file).suffix.lower()
            if ext in image_extensions:
                image_path = os.path.join(root, file)
                result = fix_image_orientation(image_path)
                if result:
                    fixed_count += 1
                elif result is False:
                    skipped_count += 1
                else:
                    error_count += 1

    return fixed_count, skipped_count, error_count

def main():
    # Get the project root (parent of scripts directory)
    script_dir = Path(__file__).parent
    project_root = script_dir.parent
    images_dir = project_root / 'public' / 'images'

    print(f"Processing images in: {images_dir}")
    print("-" * 50)

    total_fixed = 0
    total_skipped = 0
    total_errors = 0

    # Process each subdirectory
    for subdir in ['projects', 'polebarns']:
        dir_path = images_dir / subdir
        if dir_path.exists():
            print(f"\nProcessing {subdir}/...")
            fixed, skipped, errors = process_directory(dir_path)
            total_fixed += fixed
            total_skipped += skipped
            total_errors += errors

    print("\n" + "=" * 50)
    print(f"Summary:")
    print(f"  Fixed: {total_fixed}")
    print(f"  Skipped (no rotation needed): {total_skipped}")
    print(f"  Errors: {total_errors}")

if __name__ == '__main__':
    main()

File: scripts/fix-orientation.js
import sharp from 'sharp';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const imagesDir = path.join(__dirname, '..', 'public', 'images');

async function fixOrientation(inputPath) {
  try {
    const ext = path.extname(inputPath).toLowerCase();
    const tempPath = inputPath + '.tmp';

    // Read metadata first
    const metadata = await sharp(inputPath).metadata();

    // Check if image has EXIF orientation that needs fixing
    // Orientation 1 = normal, others need rotation
    if (metadata.orientation && metadata.orientation !== 1) {
      console.log(`🔄 Fixing orientation for ${path.basename(inputPath)} (orientation: ${metadata.orientation})`);

      if (ext === '.webp') {
        await sharp(inputPath)
          .rotate() // Auto-rotate based on EXIF
          .webp({ quality: 85, effort: 6 })
          .toFile(tempPath);
      } else if (['.jpg', '.jpeg'].includes(ext)) {
        await sharp(inputPath)
          .rotate() // Auto-rotate based on EXIF
          .jpeg({ quality: 90 })
          .toFile(tempPath);
      } else if (ext === '.png') {
        await sharp(inputPath)
          .rotate() // Auto-rotate based on EXIF
          .png()
          .toFile(tempPath);
      }

      // Replace original with fixed version
      await fs.promises.unlink(inputPath);
      await fs.promises.rename(tempPath, inputPath);

      console.log(`✅ Fixed ${path.basename(inputPath)}`);
      return true;
    } else {
      console.log(`⏭️  ${path.basename(inputPath)} - orientation OK`);
      return false;
    }
  } catch (error) {
    console.error(`❌ Error processing ${inputPath}:`, error.message);
    // Clean up temp file if it exists
    const tempPath = inputPath + '.tmp';
    if (fs.existsSync(tempPath)) {
      await fs.promises.unlink(tempPath);
    }
    return false;
  }
}

async function processDirectory(dir) {
  const entries = await fs.promises.readdir(dir, { withFileTypes: true });
  let fixed = 0;

  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);

    if (entry.isDirectory()) {
      fixed += await processDirectory(fullPath);
    } else if (entry.isFile()) {
      const ext = path.extname(entry.name).toLowerCase();
      if (['.jpg', '.jpeg', '.png', '.webp'].includes(ext)) {
        const wasFixed = await fixOrientation(fullPath);
        if (wasFixed) fixed++;
      }
    }
  }

  return fixed;
}

async function main() {
  console.log('🔄 Checking and fixing image orientations...\n');

  if (!fs.existsSync(imagesDir)) {
    console.error(`❌ Images directory not found: ${imagesDir}`);
    process.exit(1);
  }

  const fixed = await processDirectory(imagesDir);

  console.log(`\n✅ Done! Fixed ${fixed} image(s).`);
}

main().catch(console.error);
